src/phoenix/trace/dsl/query.py

```diff
@@ -105,61 +87,155 @@ def from_dict(cls, obj: Mapping[str, Any]) -> "Projection":
-                break
-        if not has_mapping:
-            for i, item in enumerate(seq):
-                if item is not None:
-                    yield {
-                        self.key: item,
-                        self.primary_index.key: self.primary_index(span),
-                        f"{self.position_prefix}position": i,
-                    }
-            return
-        for i, item in enumerate(seq):
-            if not isinstance(item, Mapping):
-                continue
-            record = (
-                {name: item.get(key) for name, key in self.kwargs.items()}
-                if self.kwargs
-                else dict(item)
+    def index_keys(self) -> List[str]:
+        return [self._primary_index.key, f"{self._position_prefix}position"]
+
+    def with_primary_index_key(self, _: str) -> "Explosion":
+        print("`.with_primary_index_key(...)` is deprecated and wil be removed in the future.")
+        return self
+
+    def update_sql(self, sql: Select[Any], dialect: str) -> Select[Any]:
+        array = models.Span.attributes[self.key.split(".")]
+        if dialect == _SQLITE:
+            # Because sqlite doesn't support `WITH ORDINALITY`, the order of
+            # the returned (table) values is not guaranteed. So we resort to
+            # post hoc processing using pandas.
+            return sql.where(
+                func.json_type(array) == "array",
+            ).add_columns(
+                array.label(self._array_tmp_col_label),
             )
-            for v in record.values():
-                if v is not None:
-                    break
-            else:
-                record = {}
-            if not record:
-                continue
-            record[self.primary_index.key] = self.primary_index(span)
-            record[f"{self.position_prefix}position"] = i
-            yield record
+        elif dialect == _POSTGRESQL:
+            element = (
+                func.jsonb_array_elements(array)
+                .table_valued(
+                    Column("obj", JSON),
+                    with_ordinality="position",
+                    joins_implicitly=True,
+                )
+                .render_derived()
+            )
+            obj, position = element.c.obj, element.c.position
+            return sql.where(
+                and_(
+                    func.jsonb_typeof(array) == "array",
+                    func.jsonb_typeof(obj) == "object",
+                )
+            ).add_columns(
+                # Use zero-based indexing for backward-compatibility.
+                (position - 1).label(f"{self._position_prefix}position"),
+                *(
+                    (
+                        obj[key.split(".")].label(self._add_tmp_suffix(name))
+                        for name, key in self.kwargs.items()
+                    )
+                    if self.kwargs
+                    else (obj.label(self._array_tmp_col_label),)
+                ),
```

Comment:
This is pretty hard to follow. Maybe we can break it out into a separate line?

Reply:
ok